### PR TITLE
Attribute @_implements & deriving enum equality not-named ==

### PR DESCRIPTION
**Pull request by graydon**
This implements (non-user-facing) attribute @_implements(Protocol, DeclName) that can be used to conform to a protocol requirement using a decl with a different name than the requirement.

This might sound weird, but it has (at least?) two uses we've already hit:

  1. Conforming to two protocols that use the same name for different operations (this arises when comparing floating point values)
  2. Conforming to a protocol that has a requirement that's an operator-name, without adding a new operator to the global operator list (this happens via implicit Equatable conformances on enums, and causes a lot of needless top-level inter-file dependencies)

After initial post and review, I've ticked off the following todos:
  - [x] Limit lookup to only happen when locating witnesses
  - [x] Limit lookup to the Context (protocol) name specified; currently it ignores that part of the name
  - [x] Get it working for Operator names (including protocol members)
  - [x] Add ~~~lots~~~ some more tests and fix all bugs those reveal
  - [x] Check that @_implements only mentions protocols the decl conforms to

rdar://31607695

This was originally motivated by 30959593 (operators defeat incrementality), and I've now tacked on to the end a change that @_implements the derived enum == operator in terms of a different function (creatively named `__derived_enum_equals`)

### CODE REVIEW

File: lib/AST/NameLookup.cpp [hunk 1]
```diff
@@ -1067,6 +1068,13 @@ void MemberLookupTable::addMember(Decl *member) {
   // Add this declaration to the lookup set under its compound name and simple
   // name.
   vd->getFullName().addToLookupTable(Lookup, vd);
+
+  // Add this declaration to the lookup set under any @_implements name as well
```
Comment (slavapestov) [inline lib/AST/NameLookup.cpp]: How do you prevent name lookup on concrete types from finding the decl? We only want to see the implements name while doing conformance checks no?

Comment (graydon) [inline lib/AST/NameLookup.cpp]: Correct! Forgot to put this on the todo list. Thanks!

Comment (slavapestov) [inline lib/AST/NameLookup.cpp]: I think the right place to hook this up is not inside name lookup, but in lib/Sema/TypeCheckProtocol.cpp, in particular `lookupValueWitnesses()`. When you begin conformance checking, maybe walk the members to collect all _implements attributes, building a mapping from requirement to witness inside the conformance. Then `lookupValueWitness()` can avoid most of the work by checking the pre-populated map first.

File: lib/Parse/ParseDecl.cpp [hunk 2]
```diff
@@ -437,6 +437,54 @@ bool Parser::parseSpecializeAttribute(swift::tok ClosingBrace, SourceLoc AtLoc,
...
+  if (!consumeIf(tok::r_paren, rParenLoc)) {
+    diagnose(lParenLoc, diag::attr_expected_rparen, AttrName,
+             /*DeclModifier=*/false);
+    return false;
+  }
+
+  // We parse a ParsedDeclName here just to make sure it's well formed early; we
+  // store it in the attr as a StringRef and re-parse on use, so as not to leak
+  // the ParsedDeclName type into Attr.h
+  ParsedDeclName parsedName = parseDeclName(ImplName.getValue());
```
Comment (slavapestov) [inline lib/Parse/ParseDecl.cpp]: Perhaps the semantic checking can be done in lib/Sema/TypeCheckAttr.cpp, so you're not simply throwing away the result.

**Comment by slavapestov** [comment]
I would also suggest adding some semantic checks to Sema that _implements only mentions requirements of protocols that the type conforms to.

Also do we want to allow _implements inside extensions? This might open up some interesting corner cases.

**Comment by graydon** [comment]
Pushed an update that reworks parsing and definition to handle operator names, permits the bare-minimum feature of @_implements(Equatable, ==(_:_:)). More comments welcome!

File: lib/AST/NameLookup.cpp (hunk 1 above)
Comment (graydon) [inline lib/AST/NameLookup.cpp]: Good suggestion! Fixed.

**Comment by graydon** [comment]
Pushed update to limit the name visibility to witness lookup.

**Comment by graydon** [comment]
Pushed variant that does everything on the todo list, with some new tests. Going to look at synthesizing this for the motivating case now (dependency-edge cutting) but please let me know if there's anything else you want to see for an initial merge.

**Comment by graydon** [comment]
Pushed change that derives `func==` on enums using @_implements, cutting the `provides-top-level` dependency edge.

File: lib/Sema/DerivedConformanceEquatableHashable.cpp [hunk 3]
```diff
@@ -196,7 +196,8 @@ deriveEquatable_enum_eq(TypeChecker &tc, Decl *parentDecl, EnumDecl *enumDecl) {
   //   case A, B, C
   //
   //   @derived
-  //   func ==(a: SomeEnum<T...>, b: SomeEnum<T...>) -> Bool {
+  //   @_implements(Equatable, ==(_:_:))
+  //   func _DerivedEnumEquals(a: SomeEnum<T...>, b: SomeEnum<T...>) -> Bool {
```
Comment (slavapestov) [inline lib/Sema/DerivedConformanceEquatableHashable.cpp]: Is this going to look ugly in generated interface printing? Maybe we could special case that to print this as == (or just skip synthesized decls altogether).

File: lib/Sema/TypeCheckAttr.cpp [hunk 4]
```diff
@@ -1882,6 +1884,52 @@ void AttributeChecker::visitDiscardableResultAttr(DiscardableResultAttr *attr) {
...
+  TypeResolutionOptions options;
+  options |= TR_AllowUnboundGenerics;
+
+  DeclContext *DC = D->getDeclContext();
+  Type T = TC.resolveType(ProtoTypeLoc.getTypeRepr(),
+                          DC, options);
+  ProtoTypeLoc.setType(T);
+
+  // Definite error-types were already diagnosed in resolveType.
+  if (!T || T->is<ErrorType>())
```
Comment (slavapestov) [inline lib/Sema/TypeCheckAttr.cpp]: or hasErrorType(), even

File: lib/Sema/TypeCheckProtocol.cpp [hunk 5]
```diff
@@ -1260,6 +1276,27 @@ WitnessChecker::lookupValueWitnesses(ValueDecl *req, bool *ignoringNames) {
   assert(!isa<AssociatedTypeDecl>(req) && "Not for lookup for type witnesses*");
   
   SmallVector<ValueDecl *, 4> witnesses;
+
+  // Do an initial check to see if there are any @_implements remappings
+  // for this requirement.
+  if (req->isProtocolRequirement() && req->hasName()) {
```
Comment (slavapestov) [inline lib/Sema/TypeCheckProtocol.cpp]: Maybe factor this out into a new lookupValueWitnessViaImplementsAttr() and use early return to avoid the "} cascade of death"

File: lib/Sema/TypeCheckRequest.cpp [hunk 6]
```diff
@@ -16,6 +16,7 @@
 //===----------------------------------------------------------------------===//
 #include "swift/Sema/TypeCheckRequest.h"
 #include "swift/AST/Decl.h"
+#include "swift/AST/TypeRepr.h"
```
Comment (slavapestov) [inline lib/Sema/TypeCheckRequest.cpp]: Is this necessary?

**Comment by slavapestov** [comment]
This looks great, nice job!

File: lib/Sema/DerivedConformanceEquatableHashable.cpp (hunk 3 above)
Comment (jrose-apple) [inline lib/Sema/DerivedConformanceEquatableHashable.cpp]: I think it's already marked implicit, so most people won't see it. There are just a few tests that try to print implicit declarations as well.

File: lib/Sema/TypeCheckRequest.cpp (hunk 6 above)
Comment (graydon) [inline lib/Sema/TypeCheckRequest.cpp]: Yep. the call to `std::get<0>(getTypeResolutionPayload())->getLoc()` needs it.

File: lib/Sema/TypeCheckProtocol.cpp (hunk 5 above)
Comment (graydon) [inline lib/Sema/TypeCheckProtocol.cpp]: fixed

File: lib/Sema/TypeCheckAttr.cpp (hunk 4 above)
Comment (graydon) [inline lib/Sema/TypeCheckAttr.cpp]: (Assuming you mean `hasError()`, and to replace the second disjunct with that test, then: fixed)

File: lib/Sema/DerivedConformanceEquatableHashable.cpp (hunk 3 above)
Comment (graydon) [inline lib/Sema/DerivedConformanceEquatableHashable.cpp]: Yeah, it's implicit, no printing by default.

**Comment by graydon** [comment]
Pushed variant addressing recent review comments from Slava.

**Comment by graydon** [comment]
@swift-ci please test

File: lib/Sema/DerivedConformanceEquatableHashable.cpp (hunk 3 above)
Comment (xwu) [inline lib/Sema/DerivedConformanceEquatableHashable.cpp]: Drive-by thought:

Wouldn't it be neat to have this be consistent with the "__Anonymous_field" and "__Unnamed_union" terminology introduced in #6935? (I guess it'd be something like "__Derived_conformance_enum_equals".)

Comment (jrose-apple) [inline lib/Sema/DerivedConformanceEquatableHashable.cpp]: …on that note, it's not a type, so lowercase would make more sense.

Comment (graydon) [inline lib/Sema/DerivedConformanceEquatableHashable.cpp]: Changed to `__derived_enum_equals` after checking with Jordan.

**Comment by graydon** [comment]
@swift-ci please smoke test

File: include/swift/AST/Attr.def [hunk 7]
```diff
@@ -267,6 +267,11 @@ SIMPLE_DECL_ATTR(discardableResult, DiscardableResult,
 
 SIMPLE_DECL_ATTR(GKInspectable, GKInspectable, OnVar, 66)
 
+DECL_ATTR(_implements, Implements,
+          OnFunc | OnVar | OnSubscript | OnTypeAlias
+          | NotSerialized | UserInaccessible,
```
Comment (jrose-apple) [inline include/swift/AST/Attr.def]: The attribute isn't what triggers either of those things, so we're okay.

File: lib/Sema/DerivedConformanceEquatableHashable.cpp [hunk 8]
```diff
@@ -196,7 +196,9 @@ deriveEquatable_enum_eq(TypeChecker &tc, Decl *parentDecl, EnumDecl *enumDecl) {
   //   case A, B, C
   //
   //   @derived
-  //   func ==(a: SomeEnum<T...>, b: SomeEnum<T...>) -> Bool {
+  //   @_implements(Equatable, ==(_:_:))
+  //   func __derived_conformance_enum_equals(a: SomeEnum<T...>,
```
Comment (xwu) [inline lib/Sema/DerivedConformanceEquatableHashable.cpp]: s/__derived_conformance/__derived/

Comment (graydon) [inline lib/Sema/DerivedConformanceEquatableHashable.cpp]: fixed

File: include/swift/AST/KnownIdentifiers.def [hunk 9]
```diff
@@ -90,6 +90,7 @@ IDENTIFIER_WITH_NAME(NativeClassLayout, "_NativeClass")
 // Operators
 IDENTIFIER_WITH_NAME(MatchOperator, "~=")
 IDENTIFIER_WITH_NAME(EqualsOperator, "==")
+IDENTIFIER_WITH_NAME(DerivedEnumEquals, "_DerivedEnumEquals")
```
Comment (graydon) [inline include/swift/AST/KnownIdentifiers.def]: renamed to __derived_enum_equals per conversation w/ Jordan

**Comment by graydon** [comment]
Pushed variant with fixed name and change to user-availability bit such that code completion never tells the user about `__derived_enum_equals`.

**Comment by graydon** [comment]
@swift-ci please test

**Comment by graydon** [comment]
@swift-ci please test

**Comment by graydon** [comment]
r=dgregor, merging.
